### PR TITLE
chore(build): Update `make codegen` to only run on changes

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,4 +1,7 @@
-SHELL=/bin/bash -o pipefail
+export SHELL:=/bin/bash
+export SHELLOPTS:=$(if $(SHELLOPTS),$(SHELLOPTS):)pipefail:errexit
+
+.ONESHELL:
```
Comment (alexec) [inline Makefile:4]: this allows you use `trap` in your Makefile

File: Makefile [hunk 2]
```diff
@@ -132,19 +134,11 @@ status:
...
-endif
-	@mkdir -p ui/dist
-	touch ui/dist/node_modules.marker
-
-ui/dist/app/index.html: ui/dist/node_modules.marker $(UI_FILES)
+ui/dist/app/index.html: $(UI_FILES)
 	# Build UI
 	@mkdir -p ui/dist/app
 ifeq ($(STATIC_FILES),true)
+	JOBS=max yarn --cwd ui install
```
Comment (alexec) [inline Makefile:141]: `yarn install` is fast where already done (unlike `yarn build`)

File: Makefile [hunk 3]
```diff
@@ -228,6 +222,11 @@ $(EXECUTOR_IMAGE_FILE): $(ARGOEXEC_PKGS)
 
 # generation
 
+.PHONY: codegen
+codegen: proto swagger manifests docs $(GOPATH)/bin/mockery
+	go generate ./persist/sqldb ./pkg/apiclient/workflow ./server/auth ./server/auth/sso ./workflow/executor
```
Comment (alexec) [inline Makefile:227]: `go generate ./...` took around 10s, so I made it specificy

File: Makefile [hunk 4]
```diff
@@ -236,39 +235,49 @@ $(GOPATH)/bin/mockery:
...
-codegen: status proto swagger manifests mocks docs
-
 .PHONY: crds
 crds: $(GOPATH)/bin/controller-gen
 	./hack/crdgen.sh
 
-# you cannot install a specific version using `go install`, so we do this business
-.PHONY: tools
-tools:
+$(GOPATH)/bin/controller-gen:
```
Comment (alexec) [inline Makefile:242]: I know this is verbose, but it is fast

File: Makefile [hunk 5]
```diff
@@ -438,8 +447,8 @@ pkg/apiclient/workflow/workflow.swagger.json: proto
 pkg/apiclient/workflowarchive/workflow-archive.swagger.json: proto
 pkg/apiclient/workflowtemplate/workflow-template.swagger.json: proto
 
-pkg/apiclient/_.secondary.swagger.json: hack/secondaryswaggergen.go server/static/files.go pkg/apis/workflow/v1alpha1/openapi_generated.go dist/kubernetes.swagger.json
-	go run ./hack secondaryswaggergen
+pkg/apiclient/_.secondary.swagger.json: server/static/files.go pkg/apis/workflow/v1alpha1/openapi_generated.go dist/kubernetes.swagger.json
+	go run ./hack/swagger secondaryswaggergen
```
Comment (alexec) [inline Makefile:451]: splitting the hack code up makes `go run ./hack` faster by 10s faster because it only has core deps

File: hack/crdgen.sh [hunk 6]
```diff
@@ -8,6 +8,11 @@ add_header() {
   mv tmp "$1"
 }
 
+if [ "$(ls -t manifests/base/crds/full/*.yaml | head -n1)" -nt "$(ls -t pkg/apis/workflow/v1alpha1/*.go | grep -v 'deepcopy\|generated\|test'| head -n1)" ]; then
```
Comment (alexec) [inline hack/crdgen.sh:11]: this pattern I use basically says:

if [ "the newest CRD file" is newer than "the newest source file"  ]

File: .github/workflows/ci-build.yaml [hunk 7]
```diff
@@ -114,10 +114,6 @@ jobs:
           sudo find /usr/local/include -type d | xargs sudo chmod a+rx
           rm -f $PROTOC_ZIP
           ls /usr/local/include/google/protobuf/
-      - name: Install `swagger-markdown`
```
Comment (alexec) [inline .github/workflows/ci-build.yaml:117]: I don't think this brings value

**Review by simster7 — APPROVED** [review]

**Pull request merged**